### PR TITLE
Reland deeplinking with info.plist check

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

If you look at the first commit in the PR, it's just relanding https://github.com/flutter/engine/pull/22337.  The second commit turns it off by default unless you turn it on with the Info.plist.

## Related Issues

https://github.com/flutter/flutter/issues/66294#issuecomment-696413821

## Tests

I added the following tests:

Same tests we had for https://github.com/flutter/engine/pull/22337

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#import "flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine_Test.h"
+
+FLUTTER_ASSERT_ARC
+
+@interface FlutterAppDelegateTest : XCTestCase
+@end
+
+@implementation FlutterAppDelegateTest
+
+- (void)testLaunchUrl {
```
Comment (chunhtai) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm:21]: Just curious, where do we mock FlutterDeepLinkingEnabled in this test?

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 2]
```diff
@@ -121,10 +131,43 @@ - (void)userNotificationCenter:(UNUserNotificationCenter*)center
   }
 }
 
+static BOOL IsDeepLinkingEnabled() {
+  NSDictionary* infoDict = [[NSBundle mainBundle] infoDictionary];
+  NSNumber* isEnabled = [infoDict objectForKey:@"FlutterDeepLinkingEnabled"];
```
Comment (chunhtai) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm:136]: The key name may be confusing. Developers can still do their own deeplinking if they write custom plugin. this is really for the default push route behavior.

@johnpryan any suggestion for the name?

**Review by chunhtai — COMMENTED** [review]
The logic looks good, just a concern about the naming

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm:21]: I didn't, that's why it was failing.  Fixed it.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm (hunk 2 above)
Comment (johnpryan) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: That's a lot to fit into a name, I think we'll have provide more context in our docs (either the API docs or on flutter.dev)

Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: @chunhtai sound good?  We keep it like this and define it in our docs?  I agree with john that it is impossible to convey all nuances in so few characters.

Comment (chunhtai) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: yes that sounds good to me, I have updated the android part to match

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**